keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2384 @@
+container runtime.
+
+```diff
++ // Kubernetes-managed QoS resources. These are only informational to the runtime
++ // and no CRI requests should fail because of these.
++const (
++       // QOSResourcePodQOS is the Kubernetes Pod Quality of Service Class.
++       // Possible values are "Guaranteed", "Burstable" and "BestEffort".
++       QOSResourcePodQOS = "pod-qos"
++)
+```
+
+##### Kubernetes-managed QoS-class resources
+
+It would be possible to have QoS-class resources that would be managed by
+Kubernetes/kubelet instead of the container runtime. If we specify and manage
+official well-known QoS-class resource names in the API it would be possible to
+specify Kubernetes-internal names that the container runtime would know to
+ignore (or not try to manage itself). E.g. any non-namespaced QoS-class
+resources (one without `<namespace>/` prefix in the name) would be treated as
+Kubernetes-managed and ignored by the container runtime. See [Consts](#consts)
+section below for details about QoS-class resource naming.
+
+One possible usage-scenario would be pod-level cgroup controls, e.g. cgroup v2
+memory knobs in linux (see
+[KEP-2570: Support Memory QoS with cgroups v2][kep-2570].
+
+QoS-class resources could be used to specify OOM kill behavior of individual
+containers
+(ref [PR for disabling group oom kill](https://github.com/kubernetes/kubernetes/pull/122813)).
+
+Another possible usage could be to allow the user to explicitly specify the
+desired Pod QoS class of the application (instead of implicitly deriving it
+from the resource requests/limits).
+
+##### Container-level memory QoS
+
+Container runtimes could implement an admin-configurable support for memory
+QoS, as an alternative to
+[KEP-2570: Support Memory QoS with cgroups v2][kep-2570].
+
+It would be easy to enable also for example `memory.swap.*` control knobs
+available in Linux cgroups v2 to control swap-usage on a per-container basis.
+
+##### Runtime classes
+
+The QoS-class resources mechanism could be used to replace (or
+re-implement) the current runtime classes.
+
+One benefit of using a QoS-class resource to represent runtime classes would be
+the "automatic" visibility to what runtime classes are available on each node.
+E.g. adding a new runtime class on a node by re-configuration of the runtime
+would be immediately reflected in the node status without any additional admin
+tasks (like node labeling or manual creation of RuntimeClass objects).
+
+##### Splitting Pod QoS Class
+
+Currently the Pod QoS class is an implicit property of a Pod, tied to how the
+resource requests and limits of its containers are specified. QoS-class
+resources could make it possible for users to explicitly specify the Pod QoS
+class in the Pod's `.spec`, making it possible e.g. to define a pod with a
+container whose memory limit is higher than the requests, but that is treated
+by the node as `Guaranteed`.
+
+Taking this idea further, QoS-class resources could also make it possible to
```

Comment:
This is very close to https://github.com/kubernetes/kubernetes/issues/78848.

Reply:
Thanks @kannon92 for the reference, I'll add it to the proposal. This is already speculated in [Splitting Pod QoS Class](https://github.com/kubernetes/enhancements/blob/fc082b2691fb77558983ee6ef1a9445ecf07ca85/keps/sig-node/3008-qos-class-resources/README.md#splitting-pod-qos-class) section but I wasn't aware of this open issue.